File: 백준/Silver/1874. 스택 수열/스택 수열.py
from collections import deque
import sys
input = sys.stdin.readline

n = int(input())
p = deque([]) # 만들어야하는 수열
for _ in range(n):
    p.append(int(input()))

stack = []
result = []

for number in range(1, n + 1):
    stack.append(number)
    result.append("+")

    # 스택에 넣은 수가 수열에 필요한 수면 뽑기
    while stack and stack[-1] == p[0]:
        stack.pop()
        p.popleft()
        result.append("-")

# 다 돌았는데도 완성 못했을 때 NO 출력
if stack:
    print("NO")
else:        
    for r in result:
        print(r)

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        int size = 0;
        Stack<Integer> stack = new Stack<>();

        while (n-- > 0) {
            int number = Integer.parseInt(br.readLine());

            if (number > size) {
                while (number > size) {
                    stack.push(++size);
                    sb.append("+\n");
                }
                stack.pop();
                sb.append("-\n");
            } else {
                if (stack.peek() != number) {
                    System.out.print("NO");
                    return;
                }
                stack.pop();
                sb.append("-\n");
            }
        }
        System.out.print(sb);
    }
}